ta/pkcs11/src/processing_digest.c

```diff
@@ -0,0 +1,250 @@
+pkcs2tee_algorithm(uint32_t *tee_id, struct pkcs11_attribute_head *proc_params)
+{
+	static const struct {
+		enum pkcs11_mechanism_id mech_id;
+		uint32_t tee_id;
+	} pkcs2tee_algo[] = {
+		{ PKCS11_CKM_MD5, TEE_ALG_MD5 },
+		{ PKCS11_CKM_SHA_1, TEE_ALG_SHA1 },
+		{ PKCS11_CKM_SHA224, TEE_ALG_SHA224 },
+		{ PKCS11_CKM_SHA256, TEE_ALG_SHA256 },
+		{ PKCS11_CKM_SHA384, TEE_ALG_SHA384 },
+		{ PKCS11_CKM_SHA512, TEE_ALG_SHA512 },
+	};
+	size_t n = 0;
+
+	for (n = 0; n < ARRAY_SIZE(pkcs2tee_algo); n++) {
+		if (proc_params->id == pkcs2tee_algo[n].mech_id) {
+			*tee_id = pkcs2tee_algo[n].tee_id;
+			return PKCS11_CKR_OK;
+		}
+	}
+
+	return PKCS11_RV_NOT_IMPLEMENTED;
+}
+
+static enum pkcs11_rc
+allocate_tee_operation(struct pkcs11_session *session,
+		       struct pkcs11_attribute_head *params)
+{
+	uint32_t algo = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	assert(session->processing->tee_op_handle == TEE_HANDLE_NULL);
+
+	if (pkcs2tee_algorithm(&algo, params))
+		return PKCS11_CKR_FUNCTION_FAILED;
+
+	res = TEE_AllocateOperation(&session->processing->tee_op_handle,
+				    algo, TEE_MODE_DIGEST, 0);
+	if (res)
+		EMSG("TEE_AllocateOp. failed %#"PRIx32, algo);
+
+	if (res == TEE_ERROR_NOT_SUPPORTED)
+		return PKCS11_CKR_MECHANISM_INVALID;
+
+	return tee2pkcs_error(res);
+}
+
+enum pkcs11_rc init_digest_operation(struct pkcs11_session *session,
+				     enum processing_func function __unused,
+				     struct pkcs11_attribute_head *proc_params,
+				     struct pkcs11_object *ob __unused)
+{
+	assert(processing_is_tee_digest(proc_params->id));
+
+	return allocate_tee_operation(session, proc_params);
+}
+
+/*
+ * step_digest_operation - processing digest operation step
+ *
+ * @session - current session
+ * @step - step ID in the processing (oneshot, update, final)
+ * @ptype - invocation parameter types
+ * @params - invocation parameter references
+ */
+enum pkcs11_rc step_digest_operation(struct pkcs11_session *session,
+				     enum processing_func function __unused,
+				     enum processing_step step,
+				     struct pkcs11_object *obj,
+				     uint32_t ptypes, TEE_Param *params)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	void *in_buf = NULL;
+	size_t in_size = 0;
+	void *out_buf = NULL;
+	uint32_t out_size = 0;
+	void *in2_buf = NULL;
+	uint32_t in2_size = 0;
+	void *secret_value = NULL;
+	uint32_t secret_value_size = 0;
+	bool output_data = false;
+	enum pkcs11_key_type key_type = PKCS11_CKK_UNDEFINED_ID;
+	struct active_processing *proc = session->processing;
+
+	if (TEE_PARAM_TYPE_GET(ptypes, 1) == TEE_PARAM_TYPE_MEMREF_INPUT) {
+		in_buf = params[1].memref.buffer;
+		in_size = params[1].memref.size;
+		if (in_size && !in_buf)
+			return PKCS11_CKR_ARGUMENTS_BAD;
+	}
+	if (TEE_PARAM_TYPE_GET(ptypes, 2) == TEE_PARAM_TYPE_MEMREF_INPUT) {
+		in2_buf = params[2].memref.buffer;
+		in2_size = params[2].memref.size;
+		if (in2_size && !in2_buf)
+			return PKCS11_CKR_ARGUMENTS_BAD;
+	}
```

Comment:
can drop this part, memref[2] is never used here as an input buffer.

Reply:
OK